blog/02-15-2023-build-feature-rich-persistent-dapp-on-celo-using-wagmi/build-feature-rich-persistent-dapp-on-celo-using-wagmi.md

```diff
@@ -0,0 +1,747 @@
+
+## Requirements​
+
+Install the following tools:
+
+- For an editor. I recommend VSCode.
+- NodeJs version >=14.0.0. I use version 18.12.1
+- [Setup and configure hardhat](https://docs.celo.org/blog/tutorials/advance-hardhat-configuration-on-celo-using-plugins).
+
+**What we are building**
+
+We will create a decentralized application for swapping ERC20 tokens for $Celo. For a full description, please refer to the **[readme](https://github.com/bobeu/feature-rich-persistent-dapp-on-celo-using-wagmi/README.md)**
+
+The dApp was built in two parts.
+
+1. Smart contracts
+2. Frontend
+
+**Smart contracts**
+
+- Clone project
+
+```bash
+git clone https://github.com/bobeu/feature-rich-persistent-dapp-on-celo-using-wagmi.git
+```
+- Navigate to the root folder
+
+```bash
+cd feature-rich-persistent-dapp-on-celo-using-wagmi/backend
+```
+
+- Install dependencies
+
+```bash
+yarn install
+```
+Now, let's examine the core functionalities.
+
+**Backend file/folder structure**
+
+- contracts
+- deploy
+  - 00_deploy.js
+- scripts
+  - deploy.ts
+- test
+  - swapLab-test.ts
+- _gitignore_
+- _hardhat.config.ts_
+- _package-lock.json_
+- _package.json_
+- _README.md_
+- _tsconfig.json_
+- _yarn.lock_
+
+**Contract folder**
+
+- `SwapLab.sol`
+  - For top-level ownership of the contract, we'll use the openzeppellin's _Ownable.sol_.
+
+  ```js
+  import "@openzeppelin/contracts/access/Ownable.sol";
+  ```
+
+  - _IERC20.sol_ to quickly create an ERC20 compatible token.
+  
+  ```js
+  import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+  ```
+
+  - _IERC20Metadata.sol_ is an interface for token metadata such as token names, symbols, and decimals.
+
+  ```js
+  import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+  ```
+
+  - To safely perform simple arithmetic calculations, and avoid underflow or overflow risk in contracts, we import the _SafeMath.sol_. Although this may not be necessary for recent solidity versions since they come with automated built-in checks.
+
+  ```js
+  import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+  ```
+
+  - We'll use _Address.sol_ to limit interactions to only externally owned accounts - EOA.
+
+  ```js
+  import "@openzeppelin/contracts/utils/Address.sol";
+  ```
+
+  - Since we'll be interacting with external contract addresses, to avoid a reentrancy attack, _ReentrancyGuard.sol_ will be our guard.
+
+  ```js
+  import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+  ```
+We make our contract inherit the properties and public attributes of **Ownable.sol** and **ReentrancyGuard.sol**.
+
+```js
+contract SwapLab is Ownable, ReentrancyGuard {
+  // code here
+}
+```
+
+Inside the body of the _SwapLab_ contract, we tell the compiler to use everything inside the SafeMath library for all uint256's operations. This means returns from any operation that evaluate to type `uint256` can invoke any function in the SafeMath library as the first argument.
```

Comment:
always use `` backslashes when you're referring to variables or codes in your code snippets
for example `_winning_()`